sys/include/barriers.h

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_barriers Memory Barriers
+ * @ingroup     sys
+ * @brief       Memory Barriers
+ *
+ * # FAQ
+ * ## What are Memory Barriers
+ * Memory barriers suppress specific kinds optimization performed by the
+ * compiler or (in case of out-of-order CPUs) CPUs, depending on the barrier.
+ * This will most likely result in slower program execution and, thus, is in
+ * general unwanted. However, under specific circumstances, optimization can
+ * prevent correct execution.
+ *
+ * ## Why can't I use `volatile` instead
+ * Memory accesses to `volatile` can be viewed as an alternative to
+ * @ref compiler_barrier, but only in regard to other accesses to `volatile`
+ * variables. (Some compilers do not reorder memory accesses to non-`volatile`s
+ * across accesses to `volatile`s as well, but not all compilers do so.)
+ * Resorting to mark all variables as `volatile` to enforce the correct order
```

Comment:
having one variable volatile enforces load/store to that one variable to be issued (WRT volatile rules). A memory barrier forces the compiler to forget everything it knows about all variables in the current scope. Doesn't that possibly lead to more reloads than necessary?

Reply:
At least in practical code it is observed to perform better to use barriers (see the Linux reference regarding volatile at the bottom). We need to keep in mind that the typical use case is something like:

``` C
barrier()
access_to_critical_data()
barrier()
```

It is very unlikely that memory considered to require being put behind a barrier is accessed unprotected just after the barrier.

Reply:
I totally forgot the obvious argument here:

> A memory barrier forces the compiler to forget everything it knows about all variables in the current scope.

This is an incorrect generalization. The compiler will assume that global memory is expected to be accessed during the barrier. In general local variables are still fully optimized. (When the address of the stack variables has been passed to an external function, the compiler has to treat those as potentially accessed as well.)  So the compiler will only reload globals after the barrier (and only if they are used on both sides of the barrier). Thus, the odds for the barrier to be cheaper than `volatile` are very good. (And e.g. for the Linux community this was also true in practice.)

Let me prove that GCC is able to optimize access to stack variables (unless their address is passed to external code) across barriers:

```C
#include <stdint.h>

#define memory_barrier() __sync_synchronize()

typedef struct {
	uint8_t left : 4;
	uint8_t right : 4;
} foo_t;

foo_t foo(foo_t a)
{
	a.left = 0xf;
	memory_barrier();
	a.right = 0xf;
	return a;
}
```

Output of `avr-gcc -O3 -S`:
```asm
	.file	"foo_local.c"
__SP_H__ = 0x3e
__SP_L__ = 0x3d
__SREG__ = 0x3f
__tmp_reg__ = 0
__zero_reg__ = 1
	.text
.global	foo
	.type	foo, @function
foo:
/* prologue: function */
/* frame size = 0 */
/* stack size = 0 */
.L__stack_usage = 0
	ldi r24,lo8(-1)
/* epilogue start */
	ret
	.size	foo, .-foo
	.ident	"GCC: (GNU) 8.3.0"

```